### PR TITLE
[web] Optimize BitmapCanvas. Fixes overallocation of canvas elements

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

Adds a class RenderStrategy to aid in heuristics to create optimal DOM tree+canvas when rendering. It is populated by recording canvas and used by BitmapCanvas.

- optimizes paths with rrect primitive to skip SVG generation.
- Enables drawRect/Circle/RRect with stroke (no shader) to use DOM nodes.

## Related Issues

https://github.com/flutter/flutter/issues/70346

## Tests

Covered by https://github.com/flutter/flutter/pull/70966 and golden tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: lib/web_ui/lib/src/engine/dom_canvas.dart [hunk 1]
```diff
@@ -204,18 +204,25 @@ void _applyRRectBorderRadius(html.CssStyleDeclaration style, ui.RRect rrect) {
...
+      '${_borderStrokeToCssUnit(rrect.trRadiusY)}';
+  style.borderBottomLeftRadius = '${_borderStrokeToCssUnit(rrect.blRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.blRadiusY)}';
+  style.borderBottomRightRadius = '${_borderStrokeToCssUnit(rrect.brRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.brRadiusY)}';
+}
+
+String _borderStrokeToCssUnit(double value) {
+  if (value == 0) {
+    value = 1.0;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/dom_canvas.dart:223]: Is this trying to draw a hairline? If so, does it need to be divided by DPR?

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 2]
```diff
@@ -108,15 +108,19 @@ class BitmapCanvas extends EngineCanvas {
...
+  final RenderStrategy _renderStrategy;
+
   /// Allocates a canvas with enough memory to paint a picture within the given
   /// [bounds].
   ///
   /// This canvas can be reused by pictures with different paint bounds as long
   /// as the [Rect.size] of the bounds fully fit within the size used to
   /// initialize this canvas.
-  BitmapCanvas(this._bounds, {double density = 1.0})
+  BitmapCanvas(this._bounds, RenderStrategy renderStrategy,
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/bitmap_canvas.dart:119]: There's one place that passes a pre-existing render strategy, and 70 places that create a new instance. Maybe make it optional and default to a new instance in the initializer list?

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 3]
```diff
@@ -331,19 +335,30 @@ class BitmapCanvas extends EngineCanvas {
...
+  bool _useDomForRenderingFill(SurfacePaintData paint) =>
       (_preserveImageData == false && _contains3dTransform) ||
       (_childOverdraw && _canvasPool._canvas == null &&
           paint.maskFilter == null &&
           paint.shader == null &&
           paint.style != ui.PaintingStyle.stroke);
 
+  /// Same as [_useDomForRenderingFill] but allows stroke as well.
+  ///
+  /// DOM canvas be generated for simple strokes using borders.
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:347]: Done.

File: lib/web_ui/lib/src/engine/dom_canvas.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/dom_canvas.dart:223]: Added // TODO: hairline nees to take into account both dpi and density.

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:119]: It is mostly tests. There are only 1 or 2 places in engine that create instance. Will likely make density required as well.

**Pull request merged**